saga_tools/slurm_run_python.py

```diff
@@ -0,0 +1,225 @@
+* *memory*: the amount of memory to reserve, e.g. 4G.
+* *working_directory* (optional)
+* *num_gpus* (optional, default 0): the number of GPUs to reserve.
+* *num_cpus* (optional, default 1): the number of CPUs to reserve.
+* *job_name* (optional, defaults to entry point name): the name to use for this job in SLURM.
+* *echo_template* (optional boolean, default False): whether to echo
+  the generated SLURM template (for debugging).
+* *slurm_script_path* (optional): the file to write the generated
+   SLURM batch script to (for debugging).
+
+"""
+import os
+import subprocess
+import sys
+from pathlib import Path
+from typing import Optional
+
+from attr import attrib, attrs
+from attr.validators import instance_of
+
+from immutablecollections import immutabledict
+from vistautils.memory_amount import MemoryAmount, MemoryUnit
+from vistautils.parameters import Parameters, YAMLParametersLoader
+from vistautils.range import Range
+
+import temppathlib
+
+
+def main(cluster_params: Parameters, job_param_file: Path) -> None:
+    runner = SlurmPythonRunner.from_parameters(cluster_params)
+    job_params = YAMLParametersLoader().load(job_param_file)
+    entry_point = job_params.string("entry_point")
+    memory = MemoryAmount.parse(job_params.string("memory"))
+    runner.run_entry_point(
+        entry_point_name=entry_point,
+        param_file=job_param_file,
+        partition=cluster_params.string("partition"),
+        working_directory=job_params.optional_creatable_directory("working_directory")
+        or Path(os.getcwd()),
+        num_gpus=job_params.integer("num_gpus", default=0, valid_range=Range.at_least(0)),
+        num_cpus=job_params.integer("num_cpus", default=1, valid_range=Range.at_least(1)),
+        job_name=job_params.string("job_name", default=entry_point),
+        memory_request=memory,
+        echo_template=cluster_params.boolean("echo_template", default=False),
+        slurm_script_path=job_params.optional_creatable_file("slurm_script_path"),
+    )
+
+
+@attrs(frozen=True, slots=True, kw_only=True)
+class SlurmPythonRunner:
+    conda_base_path: Path = attrib(validator=instance_of(Path))
+    conda_environment: str = attrib(validator=instance_of(str))
+    log_base_directory: Path = attrib(validator=instance_of(Path))
+
+    @staticmethod
+    def from_parameters(params: Parameters) -> "SlurmPythonRunner":
+        return SlurmPythonRunner(
+            conda_base_path=params.existing_directory("conda_base_path").absolute(),
+            conda_environment=params.string("conda_environment_name"),
+            log_base_directory=params.creatable_directory("log_directory").absolute(),
+        )
+
+    def run_entry_point(
+        self,
+        entry_point_name: str,
+        param_file: Path,
+        *,
+        partition: str,
+        working_directory: Path,
+        num_gpus: int = 0,
+        num_cpus: int = 1,
+        job_name: Optional[str] = None,
+        memory_request: MemoryAmount,
```

Comment:
This doesn't have a default argument provided so it needs to be moved up in the argument positions.

Do you want a default argument? Right now, if you don't specify `--mem`, `--mem-per-cpu`, or `--mem-per-gpu`, you get 1 GB per CPU, where a CPU is actually a core. You can see this by doing `slurmd -C` on the ADAM machine. `--mem` will be total memory per node, which in our configuration defaults to 1 task/node, and you've set `--ntasks=1`, which is fine. I don't think you need to change from using `--mem`, I'm just mentioning this so you know how memory is being allocated.

Reply:
Thanks - I think `mem` is what we want.